packages/program-boilerplate/README.md

```diff
@@ -1,17 +1,226 @@
-# SaaSquatch Program Development Library
+# SaaSquatch Program Dev Library
 
-The `program-boilerplate` package includes a variety of utility functions and boilerplate
+The
+[`program-boilerplate`](https://www.npmjs.com/package/@saasquatch/program-boilerplate)
+package includes a variety of utility functions and boilerplate
 that assist in developing SaaSquatch
 [programs](https://docs.referralsaasquatch.com/growth/quickstart/). The purpose of the
 library is to reduce code duplication and enforce consistent behavior and API across all
 programs.
 
-## Getting Started
+# Getting Started
 
-A SaaSquatch program consists of three major components:
-* General program trigger handlers ([`PROGRAM_TRIGGER`](src/types/rpc.ts#L89))
-* Program introspection handler ([`PROGRAM_INTROSPECTION`](src/types/rpc.ts#L94))
-* Program validation handlers ([`PROGRAM_VALIDATION`](src/types/rpc.ts#L103))
+Here is the basic architecture of the SaaSquatch program engine:
+
+![](media/arch.png)
+
+## Template
+Each program has a template, written in JSON. The template is often referred to as the
+"program schema", although this is technically incorrect as it is not a real schema. The
+template defines several aspects of the program, including its rules, rewards, emails,
+widgets, requirements, and more. The program template is stored in Contentful as a JSON
+field in the larger "program" content type. The whole entry including the other fields is
+known as the "program template".  Program templates are retrieved and used by the Java
+backend and are never directly touched by the program. Programs can self-modify their
+templates when activated by customers. This is called Introspection and will be discussed
+in further detail below.
```

Comment:
You could reference the program template meta schema here: https://github.com/saasquatch/schema/blob/master/src/json/ProgramTemplate.schema.json